test/e2e/network/netpol/k8s_util.go

```diff
@@ -0,0 +1,384 @@
+	pods, err := k.getPodsUncached(ns, "pod", name)
+	if err != nil {
+		return nil, err
+	}
+	if len(pods) == 0 {
+		return nil, nil
+	}
+	return &pods[0], nil
+}
+
+func (k *Kubernetes) getPodsUncached(ns string, key string, val string) ([]v1.Pod, error) {
+	v1PodList, err := k.ClientSet.CoreV1().Pods(ns).List(context.TODO(), metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("%v=%v", key, val),
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to list Pods in ns %s with key/val %s=%s", ns, key, val)
+	}
+	return v1PodList.Items, nil
+}
+
+// GetPodsFromCacheByKeyVal returns an array of all Pods in the given namespace having a k/v label pair.
+func (k *Kubernetes) GetPodsFromCacheByKeyVal(ns string, key string, val string) ([]v1.Pod, error) {
+	k.mutex.Lock()
+	p, ok := k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)]
+	k.mutex.Unlock()
+	if ok {
+		return p, nil
+	}
+
+	v1PodList, err := k.getPodsUncached(ns, key, val)
+	if err != nil {
+		return nil, err
+	}
+
+	k.mutex.Lock()
+	k.podCache[fmt.Sprintf("%v_%v_%v", ns, key, val)] = v1PodList
+	k.mutex.Unlock()
+
+	return v1PodList, nil
+}
+
+// GetPod gets a pod by namespace and name
+func (k *Kubernetes) GetPod(ns string, name string) (*v1.Pod, error) {
+	kubePod, err := k.ClientSet.CoreV1().Pods(ns).Get(context.TODO(), name, metav1.GetOptions{})
+	return kubePod, errors.Wrapf(err, "unable to get pod %s/%s", ns, name)
+}
+
+// Probe execs into a pod and checks its connectivity to another pod.
+func (k *Kubernetes) Probe(nsFrom string, podFrom string, containerFrom string, addrTo string, protocol v1.Protocol, toPort int) (bool, string, error) {
+	fromPods, err := k.GetPodsFromCacheByKeyVal(nsFrom, "pod", podFrom)
+	if err != nil {
+		return false, "", err
+	}
+	if len(fromPods) == 0 {
+		return false, "", errors.Errorf("pod %s/%s not found", nsFrom, podFrom)
+	}
+	fromPod := fromPods[0]
+
+	var cmd []string
+	switch protocol {
+	case v1.ProtocolSCTP:
+		cmd = []string{"/agnhost", "connect", fmt.Sprintf("%s:%d", addrTo, toPort), "--timeout=1s", "--protocol=sctp"}
+	case v1.ProtocolTCP:
+		cmd = []string{"/agnhost", "connect", fmt.Sprintf("%s:%d", addrTo, toPort), "--timeout=1s", "--protocol=tcp"}
+	case v1.ProtocolUDP:
+		cmd = []string{"nc", "-v", "-z", "-w", "1", "-u", addrTo, fmt.Sprintf("%d", toPort)}
+	default:
+		framework.Failf("protocol %s not supported", protocol)
+	}
+
+	commandDebugString := fmt.Sprintf("kubectl exec %s -c %s -n %s -- %s", fromPod.Name, containerFrom, fromPod.Namespace, strings.Join(cmd, " "))
+	stdout, stderr, err := k.ExecuteRemoteCommand(fromPod, containerFrom, cmd)
+	if err != nil {
+		framework.Logf("%s/%s -> %s: error when running command: err - %v /// stdout - %s /// stderr - %s", nsFrom, podFrom, addrTo, err, stdout, stderr)
+		return false, commandDebugString, nil
+	}
+	return true, commandDebugString, nil
+}
+
+// ExecuteRemoteCommand executes a remote shell command on the given pod. Will be replaced with something from framework...
```

Comment:
you have the framework ones here https://github.com/kubernetes/kubernetes/blob/3aa59f7f3077642592dc8a864fcef8ba98699894/test/e2e/framework/exec_util.go#L83-L85

Reply:
👍